extensions/amp-list/0.1/amp-list.js

```diff
@@ -340,40 +345,85 @@ export class AmpList extends AMP.BaseElement {
     );
   }
 
+  /**
+   * Returns true if element's src points to amp-state.
+   *
+   * @return {boolean}
+   * @private
+   */
+  isAmpStateSrc_() {
+    const src = this.element.getAttribute('src');
+    return (
+      isExperimentOn(this.win, 'amp-list-init-from-state') &&
+      typeof src === 'string' &&
+      startsWith(src, 'amp-state:')
+    );
+  }
+
+  /**
+   * Returns json pointed at by an amp-state protocol src. For example,
+   * src="amp-state:json.path".
+   *
+   * @return {Promise<!Object|undefined>}
+   * @private
+   */
+  getAmpStateJson_() {
+    const src = this.element.getAttribute('src');
+    const ampStatePath = src.substring('amp-state:'.length);
+    return Services.bindForDocOrNull(this.element).then(bind => {
+      if (!bind) {
+        const errorMsg =
+          'You must include amp-bind in order to use amp-state as an initial source for amp-list.';
+        user().error(TAG, errorMsg);
+        return Promise.resolve();
+      }
+      return bind.getState(ampStatePath);
+    });
+  }
+
+  /**
+   * @param {!Array|!Object|undefined} json
+   * @return {!Promise}
+   * @private
+   */
+  renderLocalData_(json) {
+    if (typeof json === 'undefined') {
+      return Promise.resolve();
+    }
+
+    const array = /** @type {!Array} */ (isArray(json) ? json : [json]);
+    this.resetIfNecessary_(/* isFetch */ false);
+    return this.scheduleRender_(array, /* append */ false);
+  }
+
   /** @override */
   mutatedAttributesCallback(mutations) {
     dev().info(TAG, 'mutate:', this.element, mutations);
     let promise;
 
-    /**
-     * @param {!Array|!Object} data
-     * @return {!Promise}
-     */
-    const renderLocalData = data => {
-      // Remove the 'src' now that local data is used to render the list.
-      this.element.setAttribute('src', '');
-      const array = /** @type {!Array} */ (isArray(data) ? data : [data]);
-      this.resetIfNecessary_(/* isFetch */ false);
-      return this.scheduleRender_(array, /* append */ false);
-    };
-
     const src = mutations['src'];
     const state = /** @type {!JsonObject} */ (mutations)['state'];
     if (src !== undefined) {
-      if (typeof src === 'string') {
+      if (this.isAmpStateSrc_()) {
+        promise = this.getAmpStateJson_().then(json =>
+          this.renderLocalData_(json)
+        );
+      } else if (typeof src === 'object' && src !== null) {
```

Comment:
Does adding the `src !== null` change behavior? 

Also can use `&& src` as a short-form if you know it's safe e.g. `src` isn't a number or empty string.

Reply:
It _could_ change behavior if somehow `src` were null, which I'm not sure is possible. It is mainly to make CC happy.

Reply:
Irrelephant now that I've restructured the code.